### DISCUSSION TITLE
Optimized replace operator

### DISCUSSION THREAD

**Discussion by dmitriyse** [dotnet/csharplang#8200] [Language Ideas]
# Current State
C# already has a ternary operator, switch expression, and null-coalescing (??) operators.
However, these constructs don't cover the following widespread use case well.
Ternary expression syntax (already works):
```CSharp
string myVariable = otherVar == string.Empty ? null: otherVar;
```
switch expression syntax (already works):
```CSharp
string myVariable = switch otherVar { string.Empty => null,  _ =>otherVar};
```

null-coalescing (??) operator, works only for nullability checks:
```CSharp
string myValue = otherVar ?? string.Empty; // No way to use different condition other than "is null"
```

# Proposal:

Optimized syntax (proposal):
```CSharp
string myVariable = otherVar |? string.Empty => null;
```
Or in general (proposal):
```CSharp
result = value |? {pattern-matching} => replacement;
```

Alternative:
```CSharp
result = value when  {pattern-matching} => replacement;
```

**Comment by huoyaoyuan** [comment]
Coalescing other values is much less common than nulls.

Moreover, comparing with null is well defined for reference types and nullable types, but for other values it may not be.

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/discussions/7880 https://github.com/dotnet/csharplang/discussions/7280 https://github.com/dotnet/csharplang/discussions/6584
